File: java/CabalAdapter.java
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * CabalAdapter — Python ↔ Java translation and Java endpoint interaction.
 *
 * Commands:
 *   py2java <file.py> [-o out.java]     Translate Python to Java via Ollama
 *   java2py <file.java> [-o out.py]     Translate Java to Python via Ollama
 *   call <url> [METHOD] [json_body]     HTTP call to a Java endpoint
 *   probe <url>                         Probe a Java endpoint for reachability and headers
 *   help                                Print this help
 *
 * Translation uses the local Ollama instance (deepseek-coder-v2).
 * HTTP operations use Java's built-in HttpClient — no external dependencies.
 */
public class CabalAdapter {

    static final String OLLAMA_URL  = System.getenv().getOrDefault("OLLAMA_BASE_URL", "http://localhost:11434") + "/api/generate";
    static final String CODER_MODEL = "deepseek-coder-v2:latest";

    static final HttpClient HTTP = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(15))
        .followRedirects(HttpClient.Redirect.NORMAL)
        .build();

    // ── Entry point ────────────────────────────────────────────────────────────

    public static void main(String[] args) throws Exception {
        if (args.length == 0) { printHelp(); System.exit(1); }
        switch (args[0]) {
            case "py2java" -> py2java(args);
            case "java2py" -> java2py(args);
            case "call"    -> call(args);
            case "probe"   -> probe(args);
            case "help", "--help", "-h" -> printHelp();
            default -> { System.err.println("Unknown command: " + args[0]); printHelp(); System.exit(1); }
        }
    }

    // ── Translation ────────────────────────────────────────────────────────────

    static void py2java(String[] args) throws Exception {
        if (args.length < 2) { System.err.println("Usage: adapter py2java <file.py> [-o out.java]"); System.exit(1); }
        String source = readFile(args[1]);
        String prompt =
            "Translate the following Python code to idiomatic Java.\n" +
            "Output ONLY the Java code in a single fenced block. No explanation.\n\n" +
            "Python:\n```python\n" + source + "\n```";
        String result = translate(prompt, "Python → Java", args[1]);
        writeOrPrint(result, args, "-o");
    }

    static void java2py(String[] args) throws Exception {
        if (args.length < 2) { System.err.println("Usage: adapter java2py <file.java> [-o out.py]"); System.exit(1); }
        String source = readFile(args[1]);
        String prompt =
            "Translate the following Java code to idiomatic Python 3.\n" +
            "Output ONLY the Python code in a single fenced block. No explanation.\n\n" +
            "Java:\n```java\n" + source + "\n```";
        String result = translate(prompt, "Java → Python", args[1]);
        writeOrPrint(result, args, "-o");
    }

    static String translate(String prompt, String label, String srcFile) throws Exception {
        System.err.println("[adapter] translating " + label + ": " + srcFile);
        System.err.println("[adapter] calling " + CODER_MODEL + " via Ollama...");
        String raw = ollamaQuery(prompt);
        return stripFence(raw);
    }

    // ── HTTP call ─────────────────────────────────────────────────────────────

    static void call(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Usage: adapter call <url> [GET|POST|PUT|DELETE|PATCH] [json_body]");
            System.exit(1);
        }
        String url    = args[1];
        String method = args.length > 2 ? args[2].toUpperCase() : "GET";
        String body   = args.length > 3 ? args[3] : "";

        System.err.println("[adapter] " + method + " " + url);

        HttpRequest.Builder builder = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .timeout(Duration.ofSeconds(30))
            .header("Content-Type", "application/json")
            .header("Accept", "application/json");

        HttpRequest request = switch (method) {
            case "POST"  -> builder.POST(HttpRequest.BodyPublishers.ofString(body)).build();
            case "PUT"   -> builder.PUT(HttpRequest.BodyPublishers.ofString(body)).build();
            case "PATCH" -> builder.method("PATCH", HttpRequest.BodyPublishers.ofString(body)).build();
            case "DELETE"-> builder.DELETE().build();
            default      -> builder.GET().build();
        };

        HttpResponse<String> resp = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println("status: " + resp.statusCode());
        resp.headers().map().forEach((k, v) -> System.out.println(k + ": " + String.join(", ", v)));
        System.out.println();
        System.out.println(resp.body());
    }

    // ── Probe ─────────────────────────────────────────────────────────────────

    static void probe(String[] args) throws Exception {
        if (args.length < 2) { System.err.println("Usage: adapter probe <url>"); System.exit(1); }
        String url = args[1];
        System.err.println("[adapter] probing " + url);

        // Try OPTIONS first (returns allowed methods), fall back to GET
        for (String method : new String[]{"OPTIONS", "GET"}) {
            try {
                HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(Duration.ofSeconds(10))
                    .method(method, HttpRequest.BodyPublishers.noBody())
                    .build();
                HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
                System.out.println("reachable : true");
                System.out.println("method    : " + method);
                System.out.println("status    : " + resp.statusCode());
                System.out.println("headers   :");
                resp.headers().map().forEach((k, v) ->
                    System.out.println("  " + k + ": " + String.join(", ", v)));
                String body = resp.body();
                if (body != null && !body.isBlank()) {
                    String snippet = body.length() > 500 ? body.substring(0, 500) + "..." : body;
                    System.out.println("body      :");
                    System.out.println("  " + snippet.replace("\n", "\n  "));
                }
                return;
            } catch (IOException e) {
                if (method.equals("GET")) {
                    System.out.println("reachable : false");
                    System.out.println("error     : " + e.getMessage());
                }
            }
        }
    }

    // ── Ollama ────────────────────────────────────────────────────────────────

    static String ollamaQuery(String prompt) throws Exception {
        String json = "{\"model\":" + jsonStr(CODER_MODEL) +
                      ",\"prompt\":" + jsonStr(prompt) +
                      ",\"stream\":false}";

        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(OLLAMA_URL))
            .timeout(Duration.ofMinutes(10))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(json))
            .build();

        HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() != 200) {
            throw new RuntimeException("Ollama error " + resp.statusCode() + ": " + resp.body());
        }

        // Extract "response" field value — Ollama returns a single JSON object
        Matcher m = Pattern.compile("\"response\":\"((?:[^\"\\\\]|\\\\.)*)\"").matcher(resp.body());
        if (!m.find()) throw new RuntimeException("Could not parse Ollama response: " + resp.body());
        return unescapeJson(m.group(1));
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    static String readFile(String path) throws IOException {
        return Files.readString(Path.of(path));
    }

    static void writeOrPrint(String content, String[] args, String flag) throws IOException {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals(flag)) {
                Path out = Path.of(args[i + 1]);
                Files.writeString(out, content);
                System.err.println("[adapter] written to " + out);
                return;
            }
        }
        System.out.println(content);
    }

    static String stripFence(String text) {
        return text.trim()
            .replaceAll("(?s)^```\\w*\\n?", "")
            .replaceAll("\\n?```\\s*$", "")
            .trim();
    }

    static String jsonStr(String s) {
        return "\"" + s
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
            + "\"";
    }

    static String unescapeJson(String s) {
        return s.replace("\\n", "\n")
                .replace("\\t", "\t")
                .replace("\\r", "\r")
                .replace("\\\"", "\"")
                .replace("\\\\", "\\");
    }

    static void printHelp() {
        System.out.println("""
            CabalAdapter — Python ↔ Java translation and Java endpoint interaction

            Usage:
              adapter py2java <file.py> [-o out.java]     Translate Python to Java
              adapter java2py <file.java> [-o out.py]     Translate Java to Python
              adapter call <url> [METHOD] [json_body]     HTTP call to a Java endpoint
              adapter probe <url>                         Probe endpoint reachability and headers
              adapter help                                Show this help

            Translation uses the local Ollama instance (deepseek-coder-v2).
            Set OLLAMA_BASE_URL to override (default: http://localhost:11434).

            Examples:
              adapter py2java scanner.py -o Scanner.java
              adapter java2py Parser.java -o parser.py
              adapter call http://localhost:8080/api/status
              adapter call http://localhost:8080/api/data POST '{"key":"value"}'
              adapter probe http://localhost:8080
            """);
    }
}

File: go/cabal_adapter.go
// CabalAdapter (Go) — Python/Java ↔ Go translation, Go execution, and HTTP endpoint interaction.
//
// Commands:
//
//	py2go   <file.py>   [-o out.go]      Translate Python to Go
//	go2py   <file.go>   [-o out.py]      Translate Go to Python
//	java2go <file.java> [-o out.go]      Translate Java to Go
//	go2java <file.go>   [-o out.java]    Translate Go to Java
//	run     <file.go>   [-- args...]     Compile and run a Go file
//	build   <file.go>   [-o binary]      Compile Go to a binary
//	call    <url>       [METHOD] [body]  HTTP call to an endpoint
//	probe   <url>                        Probe endpoint reachability and headers
//	help                                 Show this help
//
// Translation uses the local Ollama instance (deepseek-coder-v2).
// Set OLLAMA_BASE_URL to override (default: http://localhost:11434).
// Set GOBIN or GOROOT to override Go toolchain location.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

var (
	ollamaBase = envOr("OLLAMA_BASE_URL", "http://localhost:11434")
	coderModel = "deepseek-coder-v2:latest"
	httpClient = &http.Client{Timeout: 30 * time.Second}
)

func main() {
	if len(os.Args) < 2 {
		printHelp()
		os.Exit(1)
	}
	var err error
	switch os.Args[1] {
	case "py2go":
		err = translate(os.Args[2:], "Python", "Go", "py", "go")
	case "go2py":
		err = translate(os.Args[2:], "Go", "Python", "go", "py")
	case "java2go":
		err = translate(os.Args[2:], "Java", "Go", "java", "go")
	case "go2java":
		err = translate(os.Args[2:], "Go", "Java", "go", "java")
	case "run":
		err = runGo(os.Args[2:])
	case "build":
		err = buildGo(os.Args[2:])
	case "call":
		err = callEndpoint(os.Args[2:])
	case "probe":
		err = probeEndpoint(os.Args[2:])
	case "help", "--help", "-h":
		printHelp()
	default:
		fmt.Fprintf(os.Stderr, "unknown command: %s\n\n", os.Args[1])
		printHelp()
		os.Exit(1)
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, "[go-adapter] error:", err)
		os.Exit(1)
	}
}

// ── Translation ───────────────────────────────────────────────────────────────

func translate(args []string, srcLang, dstLang, srcExt, dstExt string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: go-adapter %s2%s <file.%s> [-o out.%s]",
			strings.ToLower(srcLang), strings.ToLower(dstLang), srcExt, dstExt)
	}
	src := args[0]
	code, err := os.ReadFile(src)
	if err != nil {
		return fmt.Errorf("reading %s: %w", src, err)
	}

	prompt := fmt.Sprintf(
		"Translate the following %s code to idiomatic %s.\n"+
			"Output ONLY the %s code in a single fenced block. No explanation.\n\n"+
			"%s:\n```%s\n%s\n```",
		srcLang, dstLang, dstLang, srcLang, strings.ToLower(srcLang), string(code),
	)

	fmt.Fprintf(os.Stderr, "[go-adapter] translating %s → %s: %s\n", srcLang, dstLang, src)
	fmt.Fprintf(os.Stderr, "[go-adapter] calling %s via Ollama...\n", coderModel)

	result, err := ollamaQuery(prompt)
	if err != nil {
		return err
	}
	result = stripFence(result)

	out := flagVal(args[1:], "-o")
	if out != "" {
		if err := os.WriteFile(out, []byte(result+"\n"), 0644); err != nil {
			return fmt.Errorf("writing output: %w", err)
		}
		fmt.Fprintf(os.Stderr, "[go-adapter] written to %s\n", out)
	} else {
		fmt.Println(result)
	}
	return nil
}

// ── Go execution ──────────────────────────────────────────────────────────────

func runGo(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: go-adapter run <file.go> [-- args...]")
	}
	gobin, err := findGo()
	if err != nil {
		return err
	}
	file := args[0]
	// Args after -- are passed to the program
	progArgs := []string{}
	for i, a := range args[1:] {
		if a == "--" {
			progArgs = args[i+2:]
			break
		}
	}
	fmt.Fprintf(os.Stderr, "[go-adapter] go run %s\n", file)
	cmd := exec.Command(gobin, append([]string{"run", file}, progArgs...)...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Env = parentEnv()
	return cmd.Run()
}

func buildGo(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: go-adapter build <file.go> [-o binary]")
	}
	gobin, err := findGo()
	if err != nil {
		return err
	}
	file := args[0]
	out := flagVal(args[1:], "-o")
	if out == "" {
		base := strings.TrimSuffix(filepath.Base(file), ".go")
		out = filepath.Join(".", base)
	}
	fmt.Fprintf(os.Stderr, "[go-adapter] go build %s → %s\n", file, out)
	cmd := exec.Command(gobin, "build", "-o", out, file)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Env = parentEnv()
	if err := cmd.Run(); err != nil {
		return err
	}
	fmt.Fprintf(os.Stderr, "[go-adapter] built: %s\n", out)
	return nil
}

// ── HTTP ──────────────────────────────────────────────────────────────────────

func callEndpoint(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: go-adapter call <url> [GET|POST|PUT|DELETE|PATCH] [json_body]")
	}
	url := args[0]
	method := "GET"
	body := ""
	if len(args) > 1 {
		method = strings.ToUpper(args[1])
	}
	if len(args) > 2 {
		body = args[2]
	}
	fmt.Fprintf(os.Stderr, "[go-adapter] %s %s\n", method, url)

	var bodyReader io.Reader
	if body != "" {
		bodyReader = strings.NewReader(body)
	}
	req, err := http.NewRequest(method, url, bodyReader)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	fmt.Printf("status: %d %s\n", resp.StatusCode, resp.Status)
	for k, v := range resp.Header {
		fmt.Printf("%s: %s\n", k, strings.Join(v, ", "))
	}
	fmt.Println()
	respBody, _ := io.ReadAll(resp.Body)
	fmt.Println(string(respBody))
	return nil
}

func probeEndpoint(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: go-adapter probe <url>")
	}
	url := args[0]
	fmt.Fprintf(os.Stderr, "[go-adapter] probing %s\n", url)

	client := &http.Client{Timeout: 10 * time.Second}
	for _, method := range []string{"OPTIONS", "GET"} {
		req, err := http.NewRequest(method, url, nil)
		if err != nil {
			return err
		}
		resp, err := client.Do(req)
		if err != nil {
			if method == "GET" {
				fmt.Printf("reachable : false\nerror     : %s\n", err)
				return nil
			}
			continue
		}
		defer resp.Body.Close()
		fmt.Printf("reachable : true\nmethod    : %s\nstatus    : %d %s\nheaders   :\n", method, resp.StatusCode, resp.Status)
		for k, v := range resp.Header {
			fmt.Printf("  %s: %s\n", k, strings.Join(v, ", "))
		}
		respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		if len(bytes.TrimSpace(respBody)) > 0 {
			fmt.Printf("body      :\n  %s\n", strings.ReplaceAll(string(respBody), "\n", "\n  "))
		}
		return nil
	}
	return nil
}

// ── Ollama ────────────────────────────────────────────────────────────────────

type ollamaReq struct {
	Model  string `json:"model"`
	Prompt string `json:"prompt"`
	Stream bool   `json:"stream"`
}

type ollamaResp struct {
	Response string `json:"response"`
}

func ollamaQuery(prompt string) (string, error) {
	payload, err := json.Marshal(ollamaReq{Model: coderModel, Prompt: prompt, Stream: false})
	if err != nil {
		return "", err
	}
	client := &http.Client{Timeout: 10 * time.Minute}
	resp, err := client.Post(ollamaBase+"/api/generate", "application/json", bytes.NewReader(payload))
	if err != nil {
		return "", fmt.Errorf("ollama unreachable: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("ollama error %d: %s", resp.StatusCode, body)
	}
	var result ollamaResp
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("parsing ollama response: %w", err)
	}
	return result.Response, nil
}

// ── Helpers ───────────────────────────────────────────────────────────────────

var fenceRe = regexp.MustCompile("(?s)^```\\w*\\n?|\\n?```\\s*$")

func stripFence(s string) string {
	return strings.TrimSpace(fenceRe.ReplaceAllString(strings.TrimSpace(s), ""))
}

func flagVal(args []string, flag string) string {
	for i := 0; i < len(args)-1; i++ {
		if args[i] == flag {
			return args[i+1]
		}
	}
	return ""
}

func findGo() (string, error) {
	// Respect explicit override
	if g := os.Getenv("GOBIN"); g != "" {
		return g, nil
	}
	// Search well-known install paths (not system PATH — it may be restricted)
	candidates := []string{
		"/usr/local/go/bin/go",
		"/usr/bin/go",
		"/opt/go/bin/go",
		"/snap/bin/go",
	}
	if goroot := os.Getenv("GOROOT"); goroot != "" {
		candidates = append([]string{filepath.Join(goroot, "bin", "go")}, candidates...)
	}
	for _, p := range candidates {
		if _, err := os.Stat(p); err == nil {
			return p, nil
		}
	}
	return "", fmt.Errorf("go toolchain not found; set GOROOT or GOBIN, or symlink go into tools/")
}

// parentEnv returns the process environment with PATH restored to the real
// system PATH so go run / go build can find the standard library and toolchain.
func parentEnv() []string {
	env := os.Environ()
	for i, e := range env {
		if strings.HasPrefix(e, "PATH=") {
			// Replace restricted PATH with real system fallback
			env[i] = "PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
			return env
		}
	}
	return append(env, "PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin")
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func printHelp() {
	fmt.Print(`go-adapter — Python/Java ↔ Go translation, Go execution, and endpoint interaction

Usage:
  go-adapter py2go   <file.py>   [-o out.go]      Translate Python to Go
  go-adapter go2py   <file.go>   [-o out.py]      Translate Go to Python
  go-adapter java2go <file.java> [-o out.go]      Translate Java to Go
  go-adapter go2java <file.go>   [-o out.java]    Translate Go to Java
  go-adapter run     <file.go>   [-- args...]     Compile and run a Go file
  go-adapter build   <file.go>   [-o binary]      Compile Go to a binary
  go-adapter call    <url>       [METHOD] [body]  HTTP call to an endpoint
  go-adapter probe   <url>                        Probe endpoint reachability
  go-adapter help                                 Show this help

Translation uses the local Ollama instance (deepseek-coder-v2).
  OLLAMA_BASE_URL  override Ollama endpoint  (default: http://localhost:11434)
  GOROOT / GOBIN   override Go toolchain location

Examples:
  go-adapter py2go   scanner.py   -o Scanner.go
  go-adapter java2go Parser.java  -o parser.go
  go-adapter run     exploit.go   -- --target 10.0.0.1
  go-adapter build   tool.go      -o tools/Network/tool
  go-adapter call    http://localhost:8080/api/health
  go-adapter probe   http://localhost:9090
`)
}
